haskell/test/SMT/Model/BaseDAO/Types.hs

```diff
@@ -0,0 +1,263 @@
+                      }
+                in (Map.insert addr (SimpleFA2ContractType updatedSc) contracts, tez)
+              _ ->
+                (contracts, toMutez 0)
+          OtherOperation addr param tez ->
+            case Map.lookup addr contracts of
+              Just (OtherContractType oc) ->
+                let updatedSc = oc
+                      { ocStorage = param : (oc & ocStorage)
+                      , ocMutez = unsafeAddMutez tez (oc & ocMutez)
+                      }
+                in (Map.insert addr (OtherContractType updatedSc) contracts, tez)
+              _ ->
+                (contracts, toMutez 0)
+
+  currentBal <- get <&> msMutez
+  let newBal = case currentBal `subMutez` minusBal of
+        Just b -> b
+        Nothing -> error "Dao balance is negative."
+
+  modify $ \ms -> ms { msContracts = updatedContracts, msMutez = newBal }
+
+getStore :: ModelT Storage
+getStore = get <&> msFullStorage <&> fsStorage
+
+getConfig :: ModelT Config
+getConfig = get <&> msFullStorage <&> fsConfig
+
+modifyStore :: (Storage -> ModelT Storage) -> ModelT ()
+modifyStore f = do
+  ms <- get
+  updatedStorage <- f (ms & msFullStorage & fsStorage)
+  put $ ms
+    { msFullStorage = (ms & msFullStorage)
+        { fsStorage = updatedStorage }
+    }
+
+-- Simple operation that only allows transfering to `FA2.TransferItem` entrypoint
+data SimpleOperation
+  = FA2TransferOperation Address [FA2.TransferItem] Mutez
+  | OtherOperation Address Text Mutez
+  deriving stock (Eq, Show, Generic)
+
+instance Buildable SimpleOperation where
+  build = genericF
+
+
+-- | A simple contract type used to track originate contracts
+-- We mostly care about contract that has the FA2 transfer entrypont.
+-- Other contract call params will be convert to text.
+data SimpleContractType
+  = SimpleFA2ContractType SimpleFA2Contract
+  | OtherContractType OtherContract
+  deriving stock (Eq, Show, Generic)
+
+instance Buildable SimpleContractType where
+  build = genericF
+
+data SimpleFA2Contract = SimpleFA2Contract
+  { sfcStorage :: [FA2.TransferParams]
+  , sfcMutez :: Mutez
+  } deriving stock (Eq, Show, Generic)
+
+instance Buildable SimpleFA2Contract where
+  build = genericF
+
+
+data OtherContract = OtherContract
+  { ocStorage :: [Text]
+  , ocMutez :: Mutez
+  } deriving stock (Eq, Show, Generic)
+
+instance Buildable OtherContract where
+  build = genericF
+
+-- | Definition of an entrypoint call to emulate in the SMTs
+data ModelCall = ModelCall
+  { mcAdvanceLevel :: Maybe Natural -- Advance level before calling the entrypoint
+  , mcParameter :: Parameter
+  , mcSource :: ModelSource
+  } deriving stock (Eq, Show, Generic)
+
+instance Buildable ContractHash where
+  build = genericF
+
+instance Buildable ModelCall where
+  build = genericF
+
+-- | Definition of an @source@ and @sender@ for a call to emulate in the SMTs
+data ModelSource = ModelSource
+  { msoSender :: Address
+  , msoSource :: Address
```

Comment:
What does the `msoSource` represent and used for? I see `msoSender` being used in `callLigoEntrypoint`, but I can't tell where `msoSource` is used.

Reply:
This is used as part of `applyDropProposal`: https://github.com/tqtezos/baseDAO/pull/282/files#diff-1bfcd009e236eb0184a58550a169351cb7e9cd2937a7cdf7ac42a23d9eb673ebR339

On the side note, is it possible to generate only contract address? There is `genContractAddress` in `Gen.Tezos.Address`, but unfortunately it is not imported. Would be nice to have it.

Reply:
> This is used as part of `applyDropProposal`

1. I see, it's used in the SMT model implementation (`handleCallViaHaskell`), but it's not used in `handleCallViaLigo`. Shouldn't it?
2. It seems our generator always uses the same address for the Sender and for the Source. Should we change it so that sometimes they're different?

> On the side note, is it possible to generate only contract address? There is genContractAddress in Gen.Tezos.Address, but unfortunately it is not exported. Would be nice to have it.

Good point, yeah, `Hedgehog.Gen.Tezos.Address` should also export `genKeyAddress` and `genContractAddress`.

I'll create an issue for this.

Out of curiosity, what's the use case for `genContractAddress`?

Reply:
https://gitlab.com/morley-framework/morley/-/issues/629

Reply:
> Out of curiosity, what's the use case for genContractAddress?

In our use case, we need the `guardian` address to always be a contract address. 

>  but it's not used in handleCallViaLigo. Shouldn't it?

Yeah, since `msSource` is meant to be the address from `SOURCE` instr, we would have to originate a contract with this `msSource` address, and let the sender call this contract, which then the contract will call basedao ligo. This way we would get the correct value of `SOURCE`. 

It is a bit non-trivial, I'll add this in the #295 as well.